### PR TITLE
actually use the highest std flag we find, up nvcc std version

### PR DESCRIPTION
**Pull request by trws**
This just fixes some odd inconsistencies we had in applying c++ standard
selection arguments to the compiler.  Also adds support for the -std
c++14 argument to nvcc 9.  I've tested this with nvcc 9 on clang, and
gcc-4.9 separately and both work.  Will see if they work *together* on
travis.  May need an extra check for host compiler type and/or version
before selecting that nvcc flag.

So it doesn't raise questions, the change in the base CMakeLists.txt is so that we find cuda before selecting compiler flags.  It's required to be able to check the nvcc version available.

### CODE REVIEW

**Comment by trws** [comment]
Travis appears to be using gcc 4.8 headers with g++4.9 for some reason... yay...

**Comment by trws** [comment]
As a bit of background, probably of most interest to @ajkunen and @willkill07, if this passes everywhere, and we double-check some things, I'm pretty sure all the compilers we currently support, including cuda 9 but excepting cuda 8, also support (in c++14 mode):

* decltype(auto) return types
* auto return types
* polymorphic lambdas (need to be careful not to declare __device__ lambdas inside these

If that's true, and we're careful about testing, we could do some major things with decltype(auto) that just plain can't be done without it.  Things like returning lambdas, or templates instantiated with lambdas.  Something to think about.

**Comment by willkill07** [comment]
This seems sensible to me. It cleans up the checking quite a bit (realistically, PGI could be added to the CMake known compilers list).

r.e. C++14 support: I am a personal fan and `decltype(auto)` would be life-changing for me when writing backend implementations (especially for reducers). I'm a proponent for it, but I will still avoid C++14 for the time being until the vendor compilers have it all working as it should. (I'd be surprised if XL or PGI or Intel had all C++14 support. Even GCC 4.9 doesn't have *great* support for C++14. But you are correct in identifying the features we would get.

**Comment by trws** [comment]
If we're willing to up the cmake required version, all of Intel, PGI and XL could be added to that list but none of those three were fully supported for compiler features in version 3.3 (which is what we require right now).  In fact, if we were willing to require the previous stable version, 3.8, even nvcc would be supported.

As to C++14, yeah, PGI and XL at least are certainly behind, but if some of the easier to implement banner features happen to be supported by all of our targets they may be worth using, at least under guards.  Specifically decltype(auto) would help with compile times for complex tuple ops if nothing else, and if we're willing to unguard it there are certain features that get a lot faster and easier, not to mention in some cases *possible*, to implement.

**Comment by trws** [comment]
As a data-point, the example below compiles and runs correctly with all of:

* g++ 4.9.3
* xl 2016.12.02 (have to use c++1y rather than 14, but works just fine)
* icpc 16.0.258
* pgi 16.10
* clang... everything, everything 3.4+ anyway

```c++
#include <tuple>
#include <type_traits>
#include <iostream>
#include <typeinfo>
template<typename T, size_t...Tidx, typename U, size_t...Uidx>
auto tcat14_helper(T l, std::index_sequence<Tidx...>, U r, std::index_sequence<Uidx...>) {
    return std::make_tuple(std::get<Tidx>(l)..., std::get<Uidx>(r)...);
}

template<typename T, typename U>
auto tcat14(T l, U r) {
    return tcat14_helper(l, std::make_index_sequence<std::tuple_size<T>::value>{},
            r, std::make_index_sequence<std::tuple_size<U>::value>{});
}

int main(int argc, char *argv[])
{
    auto t = tcat14(std::make_tuple(1, 5.0), std::make_tuple(0ul, 1l, 4.2f));
    std::cout << typeid(t).name() << std::endl;

    return 0;
}
```

Without deduced return types, that function takes about 150 lines of metaprogramming just to calculate the return type...

**Comment by trws** [comment]
That is a very good point @willkill07, I'll make it overridable.

File: .travis.yml [hunk 1]
```diff
@@ -96,7 +96,7 @@ script:
 - mkdir travis-build &&
   cd travis-build
 - cmake -DCMAKE_CXX_COMPILER="${COMPILER}" ${CMAKE_EXTRA_FLAGS} ../
-- make -j 3
+- make -j 3 VERBOSE=1
```
Comment (rhornung67) [inline .travis.yml]: 👍 this will help when reading through the logs

**Comment by trws** [comment]
We probably could, but I have a feeling it would be a bit nasty.  We'd have to actually run code to see what happens I think.  Alternately we could make the default to not supply one, which causes nvcc to emit some pretty portable code.

File: cmake/SetupCompilers.cmake [hunk 2]
```diff
@@ -40,11 +40,30 @@
 #
 ###############################################################################
 
-if (NOT RAJA_ENABLE_CLANG_CUDA)
-  set(CMAKE_CXX_STANDARD 14)
-else()
-  set(CMAKE_CXX_STANDARD 11)
-endif ()
+set(COMPILERS_KNOWN_TO_PALEOLITHIC_CMAKE AppleClang Clang GNU MSVC)
```
Comment (davidbeckingsale) [inline cmake/SetupCompilers.cmake]: Whilst this variable name is hillarious, something more descriptive might be better.

File: cmake/SetupCompilers.cmake [hunk 3]
```diff
@@ -40,11 +40,30 @@
...
+    foreach(flag_var "-std=c++17" "-std=c++1z" "-std=c++14" "-std=c++1y" "-std=c++11")
+      CHECK_CXX_COMPILER_FLAG(${flag_var} COMPILER_SUPPORTS_${flag_var})
+      if(COMPILER_SUPPORTS_${flag_var})
+        set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} ${flag_var}")
+        break()
+      endif()
+    endforeach(flag_var)
+  endif()
+else(RAJA_CXX_STANDARD_FLAG MATCHES default)
+  set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} ${RAJA_CXX_STANDARD_FLAG}")
```
Comment (davidbeckingsale) [inline cmake/SetupCompilers.cmake]: Might be worth printing a message like:

`Using C++ standard flag: ${RAJA_CXX_STANDARD_FLAG}`

File: cmake/SetupCompilers.cmake [hunk 4]
```diff
@@ -102,12 +102,18 @@ if ( MSVC )
...
+  endif()
   if(CMAKE_BUILD_TYPE MATCHES Release)
-    set(RAJA_NVCC_FLAGS -O2; -restrict; -arch ${RAJA_CUDA_ARCH}; -std c++11; --expt-extended-lambda; -ccbin; ${CMAKE_CXX_COMPILER} CACHE LIST "")
+    set(RAJA_NVCC_FLAGS -O2; -restrict; -arch ${RAJA_CUDA_ARCH}; -std ${RAJA_NVCC_STD}; --expt-extended-lambda; -ccbin; ${CMAKE_CXX_COMPILER} CACHE LIST "")
   elseif(CMAKE_BUILD_TYPE MATCHES RelWithDebInfo)
-    set(RAJA_NVCC_FLAGS -g; -G; -O2; -restrict; -arch ${RAJA_CUDA_ARCH}; -std c++11; --expt-extended-lambda; -ccbin ${CMAKE_CXX_COMPILER} CACHE LIST "")
-  elseif(CMAKE_BUILD_TYPE MATCHES Debug)
-    set(RAJA_NVCC_FLAGS -g; -G; -O0; -restrict; -arch ${RAJA_CUDA_ARCH}; -std c++11; --expt-extended-lambda; -ccbin ${CMAKE_CXX_COMPILER} CACHE LIST "")
+    set(RAJA_NVCC_FLAGS -g; -G; -O2; -restrict; -arch ${RAJA_CUDA_ARCH}; -std  ${RAJA_NVCC_STD}; --expt-extended-lambda; -ccbin ${CMAKE_CXX_COMPILER} CACHE LIST "")
+  else(CMAKE_BUILD_TYPE MATCHES Debug)
```
Comment (davidbeckingsale) [inline cmake/SetupCompilers.cmake]: I don't think our "else" case should be adding `-g`, `-G`, and `-O0`. It should probably just add `-std`, `-ccbin` and `--expt-extended-lambda`.

File: cmake/SetupPackages.cmake [hunk 5]
```diff
@@ -54,17 +54,12 @@ endif()
...
 
 if (RAJA_ENABLE_CUDA)
-  find_package(CUDA)
-  if(CUDA_FOUND)
-    message(STATUS "CUDA Enabled")
-    set (CUDA_NVCC_FLAGS ${RAJA_NVCC_FLAGS})
-    set (CUDA_PROPAGATE_HOST_FLAGS OFF)
-    include_directories(${CUDA_INCLUDE_DIRS})
-  endif()
+  find_package(CUDA REQUIRED)
```
Comment (davidbeckingsale) [inline cmake/SetupPackages.cmake]: Is there a reason you removed the status message?

File: cmake/SetupCompilers.cmake (hunk 2 above)
Comment (trws) [reply cmake/SetupCompilers.cmake]: Fair point, I'll name it for the version we require.

File: cmake/SetupPackages.cmake (hunk 5 above)
Comment (trws) [reply cmake/SetupPackages.cmake]: We already get "Found CUDA: <path> (found version "9.0")", so it seemed redundant.  I can re-add it if you like.

File: cmake/SetupCompilers.cmake (hunk 4 above)
Comment (trws) [reply cmake/SetupCompilers.cmake]: Fair point.  There are only four possible values here, so I've added the fourth and changed the else branch to RelWithDebInfo, which is the usual default for package managers like spack anyway.

**Comment by trws** [comment]
Review update @davidbeckingsale?

File: cmake/SetupCompilers.cmake (hunk 3 above)
Comment (davidbeckingsale) [reply cmake/SetupCompilers.cmake]: 👍

**Pull request merged**